### ISSUE TITLE
Incorrect return type for asyncio.gather?

### ISSUE THREAD

**Issue by s-kovacevic** [python/typeshed#2652]
`asyncio.gather` return type from `tasks.pyi` seem to be different from the actual runtime type.

**example.py**
```python
import asyncio
from typing import List

async def five() -> int:
    return 5

async def runner() -> List[int]:
    fives = await asyncio.gather(five(), five(), five())
    print(type(fives))  # prints `<class 'list'>`
    print(fives)  # prints `[5, 5, 5]`
    return fives

loop = asyncio.get_event_loop()
result: List[int] = loop.run_until_complete(runner())
```
```
$  python3 --version
Python 3.6.6
$  python3 example.py               
<class 'list'>
[5, 5, 5]
$  mypy --version
mypy 0.641
$ mypy example.py                  
example.py:11: error: Incompatible return value type (got "Tuple[int, int, int]", expected "List[int]")
```

By looking at `tasks.pyi` I see no overload or any case where `gather` does not return a `Tuple`.

**Comment by srittau** [comment]
Looking at the implementation of `gather()` in both Python 3.5 and 3.7, it seems to always return a list future. The problem is that there is no way to annotate it to return a list with different types at each position. Philosophically, `gather()` should return a tuple.

All fixes I can see have other downsides. `-> List[Any]` and `-> List[Union[_T1, _T2]]` etc. lose type and length information. The latter also makes it cumbersome to retrieve values in the general case, where `_T1 != _T2`.

**Comment by s-kovacevic** [comment]
Can't think of any better solution right now. I feel like it is better to lose type information than have it wrong. Were there any similar issues in the past?

**Comment by DevilXD** [comment]
I've just ran into this one, and I agree that `List[Any]` would be a better solution here. I understand that it loses typing information, but it's better than having an incorrect type altogether. It seems that MyPy agrees to this, as the current return type as reported by `reveal_type` there is `asyncio.Future[builtins.list[Any]]`.

**Comment by aawilson** [comment]
Is the base issue here that the return type of `gather` changed after Python 3.7, which means supporting the correct typing would break compatibility with those versions?

**Comment by aawilson** [comment]
Nevermind, I poked around about in tasks.pyi and I see that the problem I'm getting locally has to do with my interpreter version running LSP not being the version I expect, I see the issue.

**Comment by JelleZijlstra** [comment]
No, because TypeVarTuple can't be used to type a heterogeneous list.

**Comment by Avasam** [comment]
Did #9678 fix this? I had a comment in AutoSplit (https://github.com/Toufool/AutoSplit/blob/master/src/capture_method/__init__.py#L200) referencing this issue and since the changes in #9678 I get the correct type inference. From `list[Any]` to `list[CameraInfo]`.
